src/main.rs

```diff
@@ -58,33 +56,70 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
     ));
     discovery.write().await.start().await.unwrap();
 
-    // Initialize state sub-network service handler, if selected
-    let (state_tx, state_handler) = if trin_config.networks.iter().any(|val| val == STATE_NETWORK) {
-        let (state_tx, state_rx) = mpsc::unbounded_channel::<StateNetworkEndpoint>();
-        let state_handler = match trin_state::initialize(state_rx) {
-            Ok(val) => val,
-            Err(msg) => panic!("Error while initializing state network: {:?}", msg),
+    // Setup Overlay database
+    let db = Arc::new(setup_overlay_db(discovery.read().await.local_enr()));
+
+    debug!("Selected networks to spawn: {:?}", trin_config.networks);
+
+    // Initialize state sub-network service and event handlers, if selected
+    let (state_jsonrpc_tx, state_handler, state_event_tx, state_network_task) =
+        if trin_config.networks.iter().any(|val| val == STATE_NETWORK) {
+            let (state_jsonrpc_tx, state_jsonrpc_rx) =
+                mpsc::unbounded_channel::<StateNetworkEndpoint>();
+            let (state_event_tx, state_event_rx) = mpsc::unbounded_channel::<TalkRequest>();
+            let state_handler = match trin_state::initialize(state_jsonrpc_rx) {
+                Ok(val) => val,
+                Err(msg) => panic!("Error while initializing state network: {:?}", msg),
+            };
+
+            let state_network_task = spawn_state_network(
+                Arc::clone(&discovery),
+                Arc::clone(&db),
+                portalnet_config.clone(),
+                state_event_rx,
+            );
+
+            (
+                Some(state_jsonrpc_tx),
+                Some(state_handler),
+                Some(state_event_tx),
+                Some(state_network_task),
+            )
+        } else {
```

Comment:
Maybe everything inside the if-block could go into a method, to trim down the size of `main` a bit, and reduce the nesting?